src/test/spec/unified_runner/mod.rs

```diff
@@ -0,0 +1,368 @@
+            }
+        }
+    }
+
+    pub fn get_client(&self, id: &str) -> &ClientEntity {
+        self.entities.get(id).unwrap().as_client()
+    }
+
+    pub fn get_database(&self, id: &str) -> &Database {
+        self.entities.get(id).unwrap().as_database()
+    }
+
+    pub fn get_collection(&self, id: &str) -> &Collection {
+        self.entities.get(id).unwrap().as_collection()
+    }
+}
+
+pub async fn run_unified_format_test(test_file: TestFile) {
+    let version_matches = SPEC_VERSIONS.iter().any(|req| {
+        if req.major != test_file.schema_version.major {
+            return false;
+        }
+        if req.minor < test_file.schema_version.minor {
+            return false;
+        }
+        // patch versions do not affect the test file format
+        true
+    });
+    if !version_matches {
+        panic!(
+            "Test runner not compatible with specification version {}",
+            &test_file.schema_version
+        );
+    }
+
+    let mut test_runner = TestRunner::new().await;
+
+    if let Some(requirements) = test_file.run_on_requirements {
+        let mut can_run_on = false;
+        for requirement in requirements {
+            if requirement.can_run_on(&test_runner.internal_client).await {
+                can_run_on = true;
+            }
+        }
+        if !can_run_on {
+            println!("Client topology not compatible with test",);
+            return;
+        }
+    }
+
+    for test_case in test_file.tests {
+        if let Some(skip_reason) = test_case.skip_reason {
+            println!("Skipping {}: {}", &test_case.description, skip_reason);
+            continue;
+        }
+
+        if test_case
+            .operations
+            .iter()
+            .any(|op| SKIPPED_OPERATIONS.contains(&op.name.as_str()))
+        {
+            println!("Skipping {}", &test_case.description);
+            continue;
+        }
+
+        println!("Running {}", &test_case.description);
+
+        if let Some(requirements) = test_case.run_on_requirements {
+            let mut can_run_on = false;
+            for requirement in requirements {
+                if requirement.can_run_on(&test_runner.internal_client).await {
+                    can_run_on = true;
+                }
+            }
+            if !can_run_on {
+                println!(
+                    "{}: client topology not compatible with test",
+                    &test_case.description
+                );
+                return;
+            }
+        }
+
+        if let Some(ref initial_data) = test_file.initial_data {
+            for data in initial_data {
+                test_runner.insert_initial_data(data).await;
+            }
+        }
+
+        if let Some(ref create_entities) = test_file.create_entities {
+            test_runner.populate_entity_map(create_entities).await;
+        }
+
+        for operation in test_case.operations {
+            match operation.object {
```

Comment:
The spec was [updated](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#operation) to disallow the save/expect fields on test runner operations, so I reverted this back to using two different cases.